vulkan_generator/handle_remapper/generator.py

```diff
@@ -0,0 +1,324 @@
+
+        private_members : List[str] = []
+        public_members : List[str] = []
+        public_functions : List[str] = []
+
+        public_members.append(codegen.create_exception_declaration("InternalConsistencyException"))
+        public_members.append(codegen.create_exception_declaration("HandleCollisionException"))
+        public_members.append(codegen.create_exception_declaration("NonDispatchableHandleRedefinitionException",
+                                                                   base_class = "HandleCollisionException"))
+        public_members.append(codegen.create_exception_declaration("RemoveNonExistantHandleException"))
+        public_members.append(codegen.create_exception_declaration("RemapNonExistantHandleException"))
+
+        for handle in all_vulkan_types.handles:
+
+            private_members.append(f"""std::map<Handle, Handle> {handle_map_name(handle)};""")
+
+            if not all_vulkan_types.handles[handle].dispatchable:
+                private_members.append(f"""std::map<Handle, int> {handle_count_map_name(handle)};""")
+
+            private_members.append("")
+
+            public_functions.append(codegen.create_function_declaration(handle_add_name(handle),
+                                                                        arguments = {"captureHandle" : "Handle",
+                                                                                           "replayHandle" : "Handle"}))
+            public_functions.append(codegen.create_function_declaration(handle_remove_name(handle),
+                                                                        arguments = {"captureHandle" : "Handle"}))
+            public_functions.append(codegen.create_function_declaration(handle_remap_name(handle),
+                                                                        return_type = "Handle",
+                                                                        arguments = {"captureHandle" : "Handle"}))
+            public_functions.append("")
+
+        remapper_class_def = codegen.create_class_definition("VulkanHandleRemapper",
+                                                             public_inheritance = ["noncopyable"],
+                                                             public_functions = public_functions,
+                                                             public_members = public_members,
+                                                             private_members = private_members)
+        remapper_h.write(remapper_class_def +"\n")
+
+        remapper_h.write(dedent("""
+            }
+            }
+
+        """))
+
+def generate_handle_remapper_cpp(file_path : Path, all_vulkan_types : types.AllVulkanTypes) :
+
+    with open(file_path, "w", encoding="ascii") as remapper_cpp:
+
+        remapper_cpp.write(codegen.generated_header())
+
+        remapper_cpp.write(dedent("""
+            #include "HandleRemapper.h"
+            #include <gtest/gtest.h>
```

Comment:
I don't think we should include this in production code.

Reply:
What is "this"? I don't understand?

Reply:
Oh, sorry. I am being stupid. You meant the gtest header. Sorry you are correct!

Reply:
Done